src/datasink/kafka.rs

```diff
@@ -175,78 +187,96 @@ fn kafka_producer(
-                        log::verbose(&format!("Sending to topic: {} with id {id}", msg.topic));
-                    }
-                    match producer.send(
-                        BaseRecord::with_opaque_to(&msg.topic, id)
-                            .payload(&msg.value)
-                            .key(&msg.key),
-                    ) {
-                        Ok(()) => {}
-                        Err((m, e)) => {
-                            // TODO: There are various problems with sending here that we should
-                            // maybe try to figure out and signal in a sensible way.  For now,
-                            // drop the message, and go back to the slow loop, do not flush the
-                            // message queue.
-                            let msg = format!("Message production error {:?} {}", e, m);
-                            let _ = control_and_errors.send(Operation::MessageDeliveryError(msg));
-                            continue 'producer_loop;
-                        }
-                    }
-
-                    // Once we're sending, send everything we've got, or we may get backed up if the
-                    // production cadence is higher than the sending cadence.
-                    match incoming_message_queue.try_recv() {
-                        Err(mpsc::TryRecvError::Empty) => {
-                            break 'sender_loop;
-                        }
-                        Err(mpsc::TryRecvError::Disconnected) => {
-                            break 'producer_loop;
-                        }
-                        Ok(m) => {
-                            msg = m;
-                        }
-                    }
+fn send_messages(
+    producer: &Box<dyn SenderAdapter>,
+    control_and_errors: &channel::Sender<Operation>,
+    mut id: usize,
+    mut backlog: Vec<Msg>,
+    flushing: bool,
+    verbose: bool,
+) -> (usize, Vec<Msg>, bool) {
+    let mut i = 0;
+    'sender_loop: while i < backlog.len() {
+        let msg = &backlog[i];
+        i += 1; // Always consume it, see below.
+        id += 1; // Always give it a new ID, even if it is later dropped.
+        if verbose {
+            log::verbose(&format!("Sending to topic: {} with id {id}", msg.topic));
+        }
+        match producer.send(
+            BaseRecord::with_opaque_to(&msg.topic, id)
+                .payload(&msg.value)
+                .key(&msg.key),
+        ) {
+            Ok(()) => {}
+            Err(m) => {
+                // An error here only means that the message could not be enqueued; sending errors
+                // are discovered in the ProducerContext.  So an error here is pretty much fatal for
+                // the message, hence we drop it.
+                let msg = format!("Message production error {m}");
+                let _ = control_and_errors.send(Operation::MessageDeliveryError(msg));
+                if !flushing {
+                    break 'sender_loop;
```

Comment:
Probably should add a comment here that breaking out of the loop is optional - it's not obvious what the benefit is, and in fact, if there are "enough" bad messages then the result will be that the backlog will become longer and longer.  The test case has some comments relating to this too.  It's possible we should just fix it now.